game-core/src/main/java/swinglib/DocumentListenerBuilder.java

```diff
@@ -0,0 +1,42 @@
+package swinglib;
+
+import javax.swing.event.DocumentEvent;
+import javax.swing.event.DocumentListener;
+import javax.swing.text.JTextComponent;
+
+
+/**
+ * Helper to create a 'DocumentListener', which is a way to listen to text events on text fields
+ * and text areas. An action listener on text fields only fires on a keypressed 'enter' event and
+ * a key listener does not fire when text is copy-pasted into the field. The document listener
+ * should reliably trigger when text is changed.
+ */
+public final class DocumentListenerBuilder {
```

Comment:
Just observing that this class doesn't really follow the _Builder_ pattern, so the name is a bit misleading.  It's more of a utility class, so following your proposal for naming utilities from a few weeks ago, `DocumentListeners` may be more appropriate.  Or even `JTextComponents`, since the target of the utility is technically `JTextComponent`.

**Unrelated thought:** Following that naming convention might make more sense across the board with the `swinglib` builders anyway.  For example, to begin building a `JButton` we now call `JButtonBuilder.builder()`.  The repeated "builder" seems awkward; a new dev might thing it's returning a builder for a `JButtonBuilder`.  `JButtons.builder()` would be slightly less awkward. :smile:  Alternatively, if we want to keep `JButtonBuilder` a top-level class, `JButtonBuilder.newInstance()` is more verbose, but less awkward and follows the _Effective Java_ factory method naming conventions.

Reply:
> so following your proposal for naming utilities from a few weeks ago

Sorry for any confusion, that proposal was strictly for hamcrest matchers, it was not meant as a general naming rule for all utilities.

## 

You raise some very good points @ssoloff

In general for the swing builders, are they true builders? I think perhaps yes. A typical example should be something like `JButton.JButtonBuilder`, in this case we are wrapping a library so the builder is not an inner class.  Otherwise, we do still have a `JButtonBuilder` and the `build` method produces a `JButton`, so the naming and return value all line up with builder pattern.

Is this a factory or a builder? I was wondering a bit between that, the factory case implies many different copies, builder is constructing just one copy. Factory and Builder have similarities, to some extent largely a difference in API

Recalling from far back, the decision to go with `builder` in the end was to try and 'fake' what a native builder would look like. We would really like to have `JButton.builder()`, short of that we have instead `JButtonBuilder.builder()`.

I think there are pro's/con's to any choice as what we would like to really do is update `JButton` itself, but cannot. So anything we do will be a trade-off. Thank you for the additional suggestion, I think three options, not sure which is best:
1. `JButtonBuilder.builder()`
2. `new JButtonBuilder()`
3. `JButtonBuilder.newInstance()`


:thinking:   I am leaning towards 2, `1` has the odd double `builder()` problem, `3` emphasizes a bit too much that we are using and creating a builder object. I'm not sure if it's completely clear what the return value would be for something like:  `JButtonBuilder.newIntance().build()`

We could maybe call these objects factories, but we from the pre-amble, we do seem to be using builders. having a symmetry and seeing a `builder().build()` seems to me important for spotting the builder pattern. 

W.R.T documentListenerBuilder, it was originally intended to provide a builder API to specify behavior an any of the three events. It turns out we always want the same event and quickly it became simplified. I think in this case we actually have a Factory. 

Should we fix DocumentListener to fit builder pattern?  eg: `DocumentListenerBuilder.builder().listener(listener).builder()`

IMO the API there is not good, should just be a factory method. I'm reluctant to to outright change the naming to factory for two reasons:
1. everything else is builder
2. should users of these API need to worry about which components they build with a factory, and which with a builder? For example, if we did have the use case where we wanted an update but not removal event, and used a builder for it, would we have both a `DocumentListenerFactory` and a `DocumentListenerBuilder`?


My inclination is ultimately avoid analysis parallysis on this, the problem is not that great and it's readily fixed once we know how to fix it. We can let the problem get worse and postpone our decision of how to fix it, let the solution present itself as more time and data accumulate. It is probably worthwhile to open a ticket to consider the 3 options above fully. Going with `new JbuttonBuilder()` is kinda appealing IMO.

Reply:
> IMO the plural is another way to indicator a factory, like `Collectors` or `Executors`.

That's my general opinion, as well, but the JDK and Guava seem to take a different tack and use the pluralized name both for factories and utilities (e.g. `Collections`, `Objects` in the JDK and `Strings`, `Enums`, `MoreObjects` in Guava).  That was, I thought, your previous argument.

Obviously, all this would be moot if Java supported the equivalent of C# or Kotlin extension methods. :smile: